phoenix-core-server/src/main/java/org/apache/phoenix/hbase/index/IndexRegionObserver.java

```diff
@@ -1583,6 +1614,17 @@ private List<Mutation> generateOnDupMutations(BatchMutateContext context, Put at
       } else {
           // Base current state off of existing row
           flattenedCells = cells;
+          // store all current cells from currentDataRowState
```

Comment:
Technically, you don't need to read all the cells. You only need columns which are being updated. You have that information in the PTable object.  The PTable that you get as part of OnDupKey is a special PTable with only the subset of columns which are being updated. This could be an important optimization because the number of cells in the row could be lot compared to the number of columns which are actually being updated.

Reply:
How about a follow-up Jira for this? More Jiras in this area are expected anyway.

Reply:
I am ok with doing this in a follow-up JIRA.